lib/pry-doc/pry_ext/show_source_with_c_internals/symbol_extractor.rb

```diff
@@ -0,0 +1,115 @@
+module Pry::CInternals
+  class SymbolExtractor
+    class << self
+      attr_accessor :file_cache
+    end
+    @file_cache = {}
+
+    def initialize(ruby_source_folder)
+      @ruby_source_folder = ruby_source_folder
+    end
+
+    def extract(info)
+      if info.original_symbol.start_with?("#define")
+        extract_macro(info)
+      elsif info.original_symbol =~ /\s+(struct|enum)\s+/
+        extract_struct(info)
+      elsif info.original_symbol.start_with?("}")
+        extract_typedef_struct(info)
+      elsif info.original_symbol =~/^typedef.*;$/
+        extract_typedef_oneliner(info)
+      else
+        extract_function(info)
+      end
+    end
+
+    private
+
+    def extract_macro(info)
+      extract_code(info) do |code|
+        return code unless code.lines.last.strip.end_with?('\\')
+      end
+    end
+
+    def extract_struct(info)
+      extract_code(info) do |code|
+        return code if balanced?(code)
+      end
+    end
+
+    def extract_typedef_struct(info)
+      extract_code(info, direction: :reverse) do |code|
+        return code if balanced?(code)
+      end
+    end
+
+    def extract_typedef_oneliner(info)
+      source_file = source_from_file(info.file)
+      return source_file[info.line]
+    end
+
+    def extract_function(info)
+      source_file = source_from_file(info.file)
+      offset, start_line = 1, info.line
+
+      if !complete_function_signature?(source_file[info.line]) && function_return_type?(source_file[info.line - 1])
+        start_line = info.line - 1
+        offset += 1
+      end
+
+      offset += 1 if !source_file[info.line].strip.end_with?("{")
+
+      extract_code(info, offset: offset, start_line: start_line) do |code|
+        return code if balanced?(code)
+      end
+    end
+
+    def extract_code(info, offset: 1, start_line: info.line, direction: :forward, &block)
+      source_file = source_from_file(info.file)
+
+      code_proc = direction == :reverse ? -> { source_file[start_line - offset..info.line].join }
+                  : -> { source_file[start_line, offset].join }
+
+      loop do
+        yield code_proc.()
+        offset += 1
+      end
+    end
+
+    def complete_function_signature?(str)
+      str =~ /\w+\s*\*?\s+\w+\(/
+    end
+
+    def function_return_type?(str)
+      str.strip =~ /[\w\*]$/
+    end
+
+    def balanced?(str)
+      tokens = CodeRay.scan(str, :c).tokens.each_slice(2)
+      token_count(tokens, /{/) == token_count(tokens, /}/)
```

Comment:
Is there a necessity to use regexps? Seems like this will be faster:

```rb
token_count(tokens, '{')
# ...
tokens.count { |v| v.first.include?(token) && v.last == :operator }
```

Reply:
@kyrylo good point i agree